src/shortcuts/__tests__/undo.ts

```diff
@@ -0,0 +1,294 @@
+import undoShortcut from '../undo'
+import { initialState } from '../../util'
+import * as undoUtils from '../../util/isUndoEnabled'
+
+describe('undo shortcut', () => {
+  const event = { preventDefault: NOOP } as Event
+  const isUndoEnabled = jest.spyOn(undoUtils, 'isUndoEnabled')
+
+  it('dispatches undo action on shortcut if undo is enabled', () => {
+    // enable undo
+    isUndoEnabled.mockImplementationOnce(() => true)
+
+    const mockStore = createMockStore()
+    const store = mockStore(initialState())
+
+    executeShortcut(undoShortcut, { store, type: 'keyboard', event })
+
+    expect(store.getActions()).toEqual([{ type: 'undoAction' }])
+  })
+
+  it('does not dispatch an undo action if undo is disabled', () => {
+    // disable undo
+    isUndoEnabled.mockImplementationOnce(() => false)
+
+    const mockStore = createMockStore()
+    const store = mockStore(initialState())
+
+    executeShortcut(undoShortcut, { store, type: 'keyboard', event })
+
+    expect(store.getActions()).toEqual([])
+  })
+
+})
+
+it('undo thought change', async () => {
+
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+      - b`
+    ),
+    { type: 'setCursor', thoughtsRanked: [{ value: 'a', rank: '0' }] },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'aa',
+      oldValue: 'a',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'a', rank: 0 }]
+    },
+    { type: 'undoAction' }
+  ])
+
+  const exported = exportContext(store.getState(), [ROOT_TOKEN], 'text/plain')
+
+  const expectedOutput = `- ${ROOT_TOKEN}
+  - a
+  - b`
+
+  expect(exported).toEqual(expectedOutput)
+})
+
+it('group all navigation actions following a thought change and undo them together', async () => {
+
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+      - b
+      - c
+      - d`
+    ),
+    { type: 'cursorDown' },
+    {
+      type: 'existingThoughtChange',
+      newValue: 'aa',
+      oldValue: 'a',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'a', rank: 0 }]
+    },
+    { type: 'cursorUp' },
+    { type: 'moveThoughtDown' },
+    { type: 'setCursor', thoughtsRanked: null },
+    { type: 'cursorBack' },
+    { type: 'undoAction' }
+  ])
+
+  // restore the cursor to it's state before navigation actions
+  const { cursor } = store.getState()
+
+  expect(cursor).toMatchObject([
+    { value: 'aa' }])
+
+})
+
+it('ignore dead actions/Combine dispensible actions with the preceding patch', () => {
+  const store = createTestStore()
+
+  store.dispatch([
+    importText(RANKED_ROOT, `
+      - a
+        - b
+        - c
+        - d`
+    ),
+    { type: 'setCursor', thoughtsRanked: null },
+    {
+      type: 'existingThoughtChange',
+      context: ['a'],
+      oldValue: 'b',
+      newValue: 'bd',
+      rankInContext: 1,
+      thoughtsRanked: [{ value: 'b', rank: 1 }]
```

Comment:
Can you please explain why it's invalid?

Reply:
@anmolarora1 Sure. It's because `b` is a child of `a`. There is no thought at `[{ value: 'b', rank: 1 }]`. It should be `[{ value: 'a', rank: 0 }, { value: 'b', rank: 1 }]`.

Reply:
That's right. 
But do we need to specify the full path while we're explicitly passing the context as `['a']` at [line 117](https://github.com/cybersemics/em/blob/d2eb0b239f3e979b6838331f3656c2dfe3d72c83/src/shortcuts/__tests__/undo.ts#L117)?

Reply:
Yes, there is some redundancy there, but `existingThoughtChange` would have to be refactored to eliminate it.